File: crates/chidori-js/benchmarks/workloads/sort.js
// Array sorting with a comparator — comparator call overhead + the engine's
// sort implementation. Uses a deterministic LCG so every runtime sorts the
// same input and must agree on the checksum.
const N = 50_000;
const ROUNDS = 6;
let seed = 123456789;
function rnd() {
  // 32-bit LCG (glibc constants), kept in unsigned range.
  seed = (seed * 1103515245 + 12345) >>> 0;
  return seed;
}
let checksum = 0;
for (let r = 0; r < ROUNDS; r++) {
  const a = new Array(N);
  for (let i = 0; i < N; i++) a[i] = rnd();
  a.sort((x, y) => x - y);
  checksum = (checksum + a[0] + a[N - 1] + a[N >> 1]) >>> 0;
}
console.log("RESULT=" + checksum);

File: crates/chidori-js/benchmarks/workloads/sort.py
# Python twin of sort.js — sorting with a deterministic LCG fill.
#
# Two deliberate divergences from a naive line-by-line port, both needed to
# keep RESULT identical to the JS runtimes:
#
# - The JS LCG multiplies a 32-bit seed by 1103515245 as a *double*, which
#   overflows 2^53 and rounds before `>>> 0` truncates to uint32. Python ints
#   would compute that product exactly and drift from every JS engine, so the
#   multiply-add is done in float (IEEE double, same rounding) and int() % 2^32
#   reproduces ToUint32.
# - list.sort() takes no comparator; Python's natural int ordering is what the
#   JS `(x, y) => x - y` comparator requests. The comparator-call overhead the
#   JS side measures has no idiomatic CPython equivalent (cmp_to_key would
#   measure its wrapper, not the sort).
N = 50_000
ROUNDS = 6
seed = 123456789


def rnd():
    global seed
    seed = int(seed * 1103515245.0 + 12345.0) % 4294967296
    return seed


checksum = 0
for r in range(ROUNDS):
    a = [0] * N
    for i in range(N):
        a[i] = rnd()
    a.sort()
    checksum = (checksum + a[0] + a[N - 1] + a[N >> 1]) % 4294967296
print("RESULT=" + str(checksum))
